#164324997 Log Society Activity points

alexmochu opened the pull request:
##### What does this PR do?
Log Society Activity points

##### Description of Task to be completed?
As a user, I should be able to click on the log points button and log my Society activity points

##### Any background context you want to provide?
- As a user, I currently cannot get a response/toast message after login my society activity.
Ticket [#164397766](https://www.pivotaltracker.com/story/show/164397766)
- As a user, I currently cannot see the validation while inputting and submitting my society activity points.
Ticket [#164397809](https://www.pivotaltracker.com/story/show/164397809)
- As a user(Fellow), I should be able to see the new logged activity added on my list of activities on his/her dashboard after Log in points for a Society activity successfully.
Ticket [#164414060](https://www.pivotaltracker.com/story/show/164414060)

##### What are the relevant Pivotal Tracker stories?
[#164324997](https://www.pivotaltracker.com/story/show/164324997)

##### Screenshots?
* Here is a demo [link](https://streamable.com/013th)
* Log activity modal
<img width="1440" alt="screenshot 2019-03-05 at 23 57 59" src="https://user-images.githubusercontent.com/18735075/53837133-ede57300-3fa2-11e9-98cd-9656983bf2a0.png">

kimani-ndegwa commented on src/app/Dashboard/components/DashboardComponent.js:
```diff
@@ -42,49 +46,80 @@ export class DashboardComponent extends Component {
+    this.setState({
+      logPoints: true,
+    });
+    console.log('logpoints', logPoints);
```
What is going on here?

alexmochu replied:
@kimani-ndegwa Once a user clicks on the Log points button the openModalLoginPointsHandler function triggered which in turn sets the logPoints state to true. This is necessary for the Log points form to appear.

kimani-ndegwa commented on src/app/Dashboard/components/DashboardComponent.js:
```diff
@@ -42,49 +46,80 @@ export class DashboardComponent extends Component {
     } = this.props;
-    const { user } = this.state;
+    const { user, logPoints } = this.state;
+    console.log('before click', logPoints);
```
Bump this.

alexmochu replied:
@kimani-ndegwa I have removed the console log and all the others.

kimani-ndegwa commented on src/app/Dashboard/components/DashboardComponent.js:
```diff
@@ -42,49 +46,80 @@ export class DashboardComponent extends Component {
+                  <h3 className='user-dashboard__title'>My Activities</h3>
+                  <div>
+                    {/* <ButtonComponent className='button__add'>
+                      <span className='fa fa-plus' />
```
This PR does not seem complete. Too much commented out code and console.logs

alexmochu replied:
@kimani-ndegwa I have updated the PR and removed all the unnecessary code.

kimani-ndegwa commented on src/app/Dashboard/components/DashboardComponent.js:
```diff
@@ -42,49 +46,80 @@ export class DashboardComponent extends Component {
+                          />
+                        )
+                        : null
+                    }
```
I think we have talked about this enough times. Evaluate the JSX before the render. This becomes unmanageable

alexmochu replied:
@kimani-ndegwa Thanks for catching this. I have removed the evaluation it was not even necessary.

kimani-ndegwa commented on src/app/Dashboard/components/DashboardComponent.js:
```diff
@@ -107,5 +142,6 @@ export default connect(
   mapStateToProps,
   {
     fetchUserActivites: actions.fetchUserActivitiesRequest,
+    loadCategories: hactions.loadCategories,
```
what is `hactions`?

alexmochu replied:
@kimani-ndegwa I have removed the hactions file and transferred all its contents which loadCategories action to the actions file. I initially had two files containing actions but now they have been consolidated to one.

kimani-ndegwa commented on src/app/Home/operations/home.data.js:
```diff
@@ -1 +1,64 @@
+// import { put, takeEvery, call } from 'redux-saga/effects';
 
+// import actions from './actions';
+// import { CATEGORIES, LOG_POINTS } from './constants';
```
What is with this file?

alexmochu replied:
@kimani-ndegwa I have deleted this file and everything on the Home folder. This is because we will no longer need them.

kimani-ndegwa commented:
I believe this PR still has a lot of work to be done. I would refrain from raising PR that are too early in the stages of development that it becomes hard to get substantial reviews. Also, there are some discussions we had around code practices that I believe should be added in moving forward. Please take keen notice on this.

Chris-Maina commented on src/app/Dashboard/components/DashboardComponent.js:
```diff
@@ -42,49 +46,80 @@ export class DashboardComponent extends Component {
-                <span>Filter</span>
-                <span className='fa fa-filter' />
-              </ButtonComponent>
+        <div className='wrapper'>
```
I had refactored this to remove the `HeroComponent, SidebarComponent, NavbarComponent ` components since they are shared on all pages. Please revert back. What needs to be there begins from line `88`.

alexmochu replied:
@Chris-Maina I reverted back

Chris-Maina commented on src/app/Dashboard/components/DashboardComponent.js:
```diff
@@ -42,49 +46,80 @@ export class DashboardComponent extends Component {
+                      <span className='fa fa-plus' />
+                      <span>Log Points</span>
+                    </ButtonComponent> */}
+                    <button type='button' onClick={this.openModalLoginPointsHandler}>Log-in points</button>
```
Is it possible we use the `ButtonComponent` instead of the button? This is to make sure we use the same styles across the app.

alexmochu replied:
@Chris-Maina Yes, it is possible that is exactly what I have done after your feedback. I have also reused it on my other files.

Chris-Maina commented on src/app/Dashboard/components/LogPointsModal.js:
```diff
@@ -0,0 +1,188 @@
+              {
+                categoryOption.supportsMultipleParticipants === true
+                  ? (
+                    <Fragment>
```
I do not think it's necessary to wrap `Textfield` in a `Fragment`. If I am wrong let me know.

alexmochu replied:
You are not wrong @Chris-Maina  there was no need of using Fragment

Chris-Maina commented on src/app/Dashboard/operations/types.js:
```diff
@@ -2,8 +2,22 @@ const FETCH_USER_ACTIVITIES_REQUEST = 'FETCH_USER_ACTIVITIES_REQUEST';
 const FETCH_USER_ACTIVITIES_SUCCESS = 'FETCH_USER_ACTIVITIES_SUCCESS';
 const FETCH_USER_ACTIVITIES_ERROR = 'FETCH_USER_ACTIVITIES_ERROR';
 
+const LOG_POINTS = {
```
Let's have a common way of declaring constants. There is no need of declaring an object to define your constants. This has been done on the line `17`

alexmochu replied:
@Chris-Maina I think declaring constants reduces repetition in certain words. It also makes them organized and still do the same work.

Kachulio1 replied:
@alexmochu  I think @Chris-Maina is saying have a common way of declaring constants e.g on line 1
and there is no need for you to create another object because your action type will look like this `types.LOG_POINTS.POST_SUCCESS`  instead of `LOG_POINTS_SUCCESS`

alexmochu replied:
@Kachulio1 @Chris-Maina  Thank you for the feedback. I have reverted to the recommended way of declaring constants

Chris-Maina commented on src/app/Dashboard/operations/logPoints.data.js:
```diff
@@ -0,0 +1,88 @@
+import actions from './actions';
+import types from './types';
+
+const fetchCategoriesUrl = 'https://private-anon-6062be0d7e-andelasocietiesapi.apiary-mock.com/api/v1/activity-types';
```
Let's use constants defined in `config` file.

alexmochu replied:
@Chris-Maina I have removed it and used the constants in the config file

alexmochu commented:
> I believe this PR still has a lot of work to be done. I would refrain from raising PRs that are too early in the stages of development that it becomes hard to get substantial reviews. Also, there are some discussions we had around code practices that I believe should be added in moving forward. Please take keen notice on this.

@kimani-ndegwa I apologise for raising the PR too early with incomplete work. I have noted this and will do my best not repeat in the future. I will also try best to ensure that my work follows the set code standards and best practices. 
This PR in now ready to be reviewed and perhaps merged.

Chris-Maina commented on src/app/Dashboard/components/tests/LogPointsModal.test.js:
```diff
@@ -0,0 +1,36 @@
+    };
+  };
+
+  it('should submit the log society activity points data', () => {
```
You missed the `expect block` in this and test case. Also, the one that follows.

alexmochu replied:
@Chris-Maina Thank you for catching that. I have added the except block

Chris-Maina commented:
Also in the video you sent, after logging an activity successfully it should show in the table of `My Activities`. If I missed that part let me know

alexmochu commented:
> Also in the video you sent, after logging an activity successfully it should show in the table of `My Activities`. If I missed that part let me know

@Chris-Maina  There is a story for that here [164414060](https://www.pivotaltracker.com/story/show/164414060)

Kachulio1 commented on package.json:
```diff
@@ -49,6 +49,7 @@
     "file-loader": "^3.0.1",
     "html-webpack-plugin": "^3.2.0",
     "jest": "^23.6.0",
+    "jquery": "^3.3.1",
```
Why do we need jquery?

alexmochu replied:
@Kachulio1 We don't need it I will remove it.

Kachulio1 commented on src/app/Dashboard/components/DashboardContainer.js:
```diff
@@ -5,6 +5,7 @@ import PropTypes from 'prop-types';
 import { ButtonComponent } from '../../common/components';
 import MyStatsComponent from './MyStatsComponent';
 import SocietyStatsComponent from './SocietyStatsComponent';
+import LogACtivityForm from './LogPointsModal';
```
The name of this import is a bit misleading importing a form from a modal file

alexmochu replied:
@Kachulio1 I have used a more relevant name for the modal

Kachulio1 commented on src/app/Dashboard/components/DashboardContainer.js:
```diff
@@ -45,22 +48,38 @@ export class DashboardContainer extends Component {
+    loadCategories();
+  }
+
+  openModalLoginPointsHandler = () => {
```
This function does the same thing as the function on line 70, I think you can have one function it will also help you have an easier name `openModalLoginPointsHandler` hehe

alexmochu replied:
@Kachulio1 I have refactored this. I only have one function for this.

Kachulio1 commented on src/app/Dashboard/components/LogPointsModal.js:
```diff
@@ -0,0 +1,168 @@
+
+/**
+   * @name LogActivityForm
+   * @summary Returns Form
```
The summery here does not explain what this component does?

alexmochu replied:
@Kachulio1 I have added a more descriptive summary

Kachulio1 commented on src/app/Dashboard/components/DashboardContainer.js:
```diff
@@ -80,8 +99,17 @@ export class DashboardContainer extends Component {
           <div className='user-dashboard__actions col-sm-12'>
             <h3 className='user-dashboard__title'>My Activities</h3>
+            <LogACtivityForm
+              className='modal'
```
What's this prop for?

alexmochu replied:
@Kachulio1 The props found in this component are used to open and close the LogPointsModal

Kachulio1 replied:
I'm asking this className='modal'

Kachulio1 replied:
I'm asking this className='modal'

alexmochu replied:
@Kachulio1 That className was doing nothing and unnecessary doing nothing

Kachulio1 commented on src/app/Dashboard/components/LogPointsModal.js:
```diff
@@ -0,0 +1,168 @@
+    close: PropTypes.func,
+  };
+
+  handleChange = prop => (event) => {
```
It's difficult for someone to know what this prop is

alexmochu replied:
@Kachulio1 I have included a summary that explains what it does.

Kachulio1 commented:
Some error in the console 

<img width="1680" alt="Screen Shot 2019-03-15 at 12 04 13" src="https://user-images.githubusercontent.com/22243090/54420548-4e9b4b00-471b-11e9-9516-3c150cabd8f2.png">

alexmochu commented:
> Some error in the console
> 
> <img alt="Screen Shot 2019-03-15 at 12 04 13" width="1680" src="https://user-images.githubusercontent.com/22243090/54420548-4e9b4b00-471b-11e9-9516-3c150cabd8f2.png">

@Kachulio1 This is a prop validation error that is being caused by the `Category` Textfield. It is being passed a Category object from the Category list array but the `Textfield` expects the value to be a string by default. Taking into consideration that there are other 3 stories on PT which highly depend on this PR to be merged first. I recommend fixing this bug through a separate story since fixing it on this PR would require me to take more time and also the other stories will have to wait.
This is the recommended Bug fix PT story [#164656891](https://www.pivotaltracker.com/story/show/164656891)

Kachulio1 commented:
Ok, fix the conflicts, one more thing should I be able to log in points or for now there is no endpoint?

alexmochu commented:
> Ok, fix the conflicts, one more thing should I be able to log in points or for now there is no endpoint?

@Kachulio1 I have fixed the conflict and also updated the PR. Now once a fellow logs an activity successfully, it is added on top of the logged activities.

Kachulio1 commented:
I really can see the logged  activity

alexmochu commented:
> I really can't see the logged activity

@Kachulio1 The bug that was causing the logged activity has been fixed. Apiary which is being used to mock the Andela Societies api was sending back a response with a wrong date format. Please confirm it is working on your side.

alexmochu commented on src/app/Dashboard/components/DashboardContainer.js:
```diff
@@ -45,23 +48,43 @@ export class DashboardContainer extends Component {
+      );
+    }
     if (loading) {
       dashboardHtml = <p>Loading ...</p>;
```
Yes, we should. Having the text `Loading` alone does create a good experience for the user. But since it was not part of this scope I can create a story for this on PT and handle it separately.

Chris-Maina replied:
I have created a Loading component @engineer-kndegwa . It is in  this [PR](https://github.com/andela/andela-societies-frontend/pull/198)
